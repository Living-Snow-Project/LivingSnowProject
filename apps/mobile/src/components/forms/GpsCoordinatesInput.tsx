import React, { useEffect, useRef, useState } from "react";
import {
  Alert,
  Pressable,
  ReturnKeyTypeOptions,
  Text,
  TextInput,
  View,
} from "react-native";
import {
  Accuracy,
  requestForegroundPermissionsAsync,
  watchPositionAsync,
} from "expo-location";
import Logger from "@livingsnow/logger";
import { formInputStyles } from "../../styles/FormInput";
import { getAppSettings } from "../../../AppSettings";
import { TestIds } from "../../constants/TestIds";
import { Placeholders } from "../../constants/Strings";

type GpsCoordinates = {
  latitude: number;
  longitude: number;
};

type GpsCoordinatesInputProps = {
  setGpsCoordinates: (latitude: number, longitude: number) => void;
  onSubmitEditing: () => void;
  coordinates?: GpsCoordinates;
};

export function GpsCoordinatesInput({
  setGpsCoordinates,
  onSubmitEditing,
  coordinates = { latitude: 0, longitude: 0 },
}: GpsCoordinatesInputProps) {
  const watchPosition = useRef<null | { remove(): void }>(null);
  const gpsCoordinatesRef = useRef<TextInput>(null);
  const [placeholderText, setPlaceholderText] = useState(
    Placeholders.GPS.AcquiringLocation
  );
  const useGps = coordinates.latitude === 0 && coordinates.longitude === 0;
  const [gpsCoordinateString, setGpsCoordinateString] = useState<string>(
    useGps ? `` : `${coordinates.latitude}, ${coordinates.longitude}`
  );
  const [manualGpsCoordinates, setManualGpsCoordinates] = useState(!useGps);
  const showGpsWarning = useGps ? getAppSettings().showGpsWarning : false;
  const clipCoordinate = (coordinate: number) =>
    JSON.stringify(coordinate.toFixed(6)).replace('"', "").replace('"', "");

  // accepts inputs from both location subscription and typed user input
  const updateGpsCoordinateString = (value: string) => {
    setGpsCoordinateString(value);

    // some users are adding parenthesis when manually entering coordinates
    // we don't want them in the data set so quietly remove
    const coords = value.replace(`(`, ``).replace(`)`, ``).split(",");

    let latitude;
    let longitude;

    if (coords[0]) {
      latitude = Number(coords[0].trim());
    }

    if (coords[1]) {
      longitude = Number(coords[1].trim());
    }

    setGpsCoordinates(latitude, longitude);
  };

  const stopGps = () => {
    if (watchPosition.current) {
      watchPosition.current?.remove();
      watchPosition.current = null;
    }
  };

  // location subscription
  useEffect(() => {
    if (!useGps) return () => {};

    (async () => {
      const { status } = await requestForegroundPermissionsAsync();
      if (status !== "granted") {
        setManualGpsCoordinates(true);
        setPlaceholderText(Placeholders.GPS.NoPermissions);
        Logger.Warn(`Permission to access foreground location was ${status}.`);
        return;
      }

      try {
        watchPosition.current = await watchPositionAsync(
          {
            accuracy: Accuracy.High,
            timeInterval: 5000,
          },
          ({ coords }) =>
            updateGpsCoordinateString(
              `${clipCoordinate(coords.latitude)}, ${clipCoordinate(
                coords.longitude
              )}`
            )
        );
      } catch (error) {
        setManualGpsCoordinates(true);
        setPlaceholderText(Placeholders.GPS.NoLocation);
        Logger.Error(`Location.watchPositionAsync() failed: ${error}`);
      }
    })();

    return stopGps;
  }, []);

  useEffect(() => {
    if (useGps) {
      gpsCoordinatesRef?.current?.focus();
    }
  }, [manualGpsCoordinates]);

  const enableManualGpsCoordinates = () => {
    stopGps();
    updateGpsCoordinateString("");
    setManualGpsCoordinates(true);
    setPlaceholderText(Placeholders.GPS.EnterCoordinates);
  };

  const confirmManualGpsCoordinates = () => {
    if (showGpsWarning) {
      Alert.alert(
        "Enter GPS coordinates manually?",
        "This message can be disabled in Settings.",
        [
          {
            text: "Yes",
            onPress: () => enableManualGpsCoordinates(),
          },
          {
            text: "No",
            style: "cancel",
          },
        ]
      );
    } else {
      enableManualGpsCoordinates();
    }
  };

  const gpsFieldProps = {
    style: formInputStyles.optionInputText,
    editable: manualGpsCoordinates,
    maxLength: 30,
    placeholder: placeholderText,
    defaultValue: gpsCoordinateString,
    returnKeyType: "done" as ReturnKeyTypeOptions,
  };

  return (
    <>
      <Text style={formInputStyles.optionStaticText}>
        GPS Coordinates (latitude, longitude)
      </Text>
      {!manualGpsCoordinates && (
        /* user confirms they want to enter GPS coordinates manually */
        <Pressable
          onPress={() => confirmManualGpsCoordinates()}
          testID={TestIds.GPS.gpsManualPressableTestId}
        >
          <View pointerEvents="none">
            <TextInput {...gpsFieldProps} />
          </View>
        </Pressable>
      )}
      {manualGpsCoordinates && (
        <TextInput
          {...gpsFieldProps}
          testID={TestIds.GPS.gpsManualInputTestId}
          onChangeText={(value) => updateGpsCoordinateString(value)}
          onSubmitEditing={() => onSubmitEditing()}
          ref={gpsCoordinatesRef}
        />
      )}
    </>
  );
}
