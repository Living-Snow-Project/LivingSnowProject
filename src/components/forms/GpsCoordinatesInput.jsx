import React, { useEffect, useRef, useState } from "react";
import {
  Alert,
  Platform,
  Text,
  TextInput,
  TouchableHighlight,
} from "react-native";
import PropTypes from "prop-types";
import * as Location from "expo-location";
import { formInputStyles } from "../../styles/FormInput";

const GpsCoordinatesInput = ({ setGpsCoordinates, onSubmitEditing }) => {
  const watchPosition = useRef(null);
  const gpsCoordinatesRef = useRef(null);
  const [gpsCoordinateString, setGpsCoordinateString] = useState(null);
  const [manualGpsCoordinates, setManualGpsCoordinates] = useState(false);
  // TextInput behaves differently on iOS and Android, detailed explaination in rendering of component
  const [gpsCoordinatesEditable, setGpsCoordinatesEditable] = useState(
    Platform.OS === "ios" ? true : false
  );

  // location subscription
  useEffect(() => {
    (async () => {
      const { status } = await Location.requestForegroundPermissionsAsync();
      if (status !== "granted") {
        console.log(
          "Permission to access foreground location was denied: ${status}"
        );
        return;
      }

      try {
        watchPosition.current = await Location.watchPositionAsync(
          {
            accuracy: Location.Accuracy.High,
            timeInterval: 5000,
          },
          ({ coords }) =>
            updateGpsCoordinateString(
              `${clipCoordinate(coords.latitude)}, ${clipCoordinate(
                coords.longitude
              )}`
            )
        );
      } catch (error) {
        console.log(
          `Location.watchPositionAsync() failed: ${JSON.stringify(error)}`
        );
      }
    })();

    return stopGps;
  }, []);

  const enableManualGpsCoordinates = () => {
    stopGps();

    // reset coordinate string, tell component user will enter coordinates, make the coordinate input editable, set focus
    updateGpsCoordinateString("");
    setManualGpsCoordinates(true);
    setGpsCoordinatesEditable(true);
    gpsCoordinatesRef.current.focus();
  };

  const clipCoordinate = (coordinate) =>
    JSON.stringify(coordinate.toFixed(6)).replace('"', "").replace('"', "");

  // accepts inputs from both location subscription and typed user input
  const updateGpsCoordinateString = (value) => {
    setGpsCoordinateString(value);

    // some users are adding parenthesis when manually entering coordinates, and we don't want them in the data set, quietly remove
    value = value.replace(`(`, ``).replace(`)`, ``);

    let latitude = undefined;
    let longitude = undefined;
    const coordinates = value.split(",");

    if (coordinates[0]) {
      latitude = coordinates[0].trim();
    }

    if (coordinates[1]) {
      longitude = coordinates[1].trim();
    }

    setGpsCoordinates(latitude, longitude);
  };

  const confirmManualGpsCoordinates = () => {
    if (global.appConfig.showGpsWarning && !manualGpsCoordinates) {
      Alert.alert(
        "Enter GPS coordinates manually?\nThis message can be disabled in Settings.",
        null,
        [
          {
            text: "Yes",
            onPress: () => enableManualGpsCoordinates(),
          },
          {
            text: "No",
            style: "cancel",
          },
        ]
      );
    } else if (!manualGpsCoordinates) {
      enableManualGpsCoordinates();
    }
  };

  const stopGps = () => {
    if (watchPosition.current) {
      watchPosition.current.remove();
      watchPosition.current = null;
    }
  };

  const gpsFieldProps = {
    style: formInputStyles.optionInputText,
    defaultValue: gpsCoordinateString,
    onChangeText: (value) => updateGpsCoordinateString(value),
    onSubmitEditing: () => onSubmitEditing(),
    ref: gpsCoordinatesRef,
    maxLength: 30,
    returnKeyType: "done",
    editable: gpsCoordinatesEditable,
  };

  return (
    <>
      {/* on iOS, TextInput captures input over parent TouchableHighlight and focus can be kept off even if editable = true.
      on Android, onTouchStart does nothing (can't make it editable through onTouchStart) and focus can't be kept off
      So use TouchableHighlight on Android for allowing manual GPS coordinates and TextInput on iOS. */}
      <Text style={formInputStyles.optionStaticText}>
        GPS Coordinates (latitude, longitude)
      </Text>
      {Platform.OS === "ios" && (
        <TextInput
          {...gpsFieldProps}
          onTouchStart={() => confirmManualGpsCoordinates()}
        />
      )}

      {Platform.OS === "android" && (
        <TouchableHighlight onPress={() => confirmManualGpsCoordinates()}>
          <TextInput {...gpsFieldProps} />
        </TouchableHighlight>
      )}
    </>
  );
};

GpsCoordinatesInput.propTypes = {
  setGpsCoordinates: PropTypes.func.isRequired,
  onSubmitEditing: PropTypes.func.isRequired,
};

export default GpsCoordinatesInput;
