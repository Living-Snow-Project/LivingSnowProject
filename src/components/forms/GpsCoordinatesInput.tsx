import React, { useEffect, useRef, useState } from "react";
import {
  Alert,
  Pressable,
  ReturnKeyTypeOptions,
  Text,
  TextInput,
  View,
} from "react-native";
import PropTypes from "prop-types";
import * as Location from "expo-location";
import { formInputStyles } from "../../styles/FormInput";
import Logger from "../../lib/Logger";
import { getAppSettings } from "../../../AppSettings";
import TestIds from "../../constants/TestIds";
import Placeholders from "../../constants/Strings";

function GpsCoordinatesInput({ setGpsCoordinates, onSubmitEditing }) {
  const watchPosition = useRef<null | { remove(): void }>(null);
  const gpsCoordinatesRef = useRef<TextInput>(null);
  const [placeholderText, setPlaceholderText] = useState(
    Placeholders.GPS.AcquiringLocation
  );
  const [gpsCoordinateString, setGpsCoordinateString] = useState<string>(``);
  const [manualGpsCoordinates, setManualGpsCoordinates] = useState(false);
  const { showGpsWarning } = getAppSettings();
  const clipCoordinate = (coordinate) =>
    JSON.stringify(coordinate.toFixed(6)).replace('"', "").replace('"', "");

  // accepts inputs from both location subscription and typed user input
  const updateGpsCoordinateString = (value) => {
    setGpsCoordinateString(value);

    // some users are adding parenthesis when manually entering coordinates
    // we don't want them in the data set so quietly remove
    const coordinates = value.replace(`(`, ``).replace(`)`, ``).split(",");

    let latitude;
    let longitude;

    if (coordinates[0]) {
      latitude = coordinates[0].trim();
    }

    if (coordinates[1]) {
      longitude = coordinates[1].trim();
    }

    setGpsCoordinates(latitude, longitude);
  };

  const stopGps = () => {
    if (watchPosition.current) {
      watchPosition.current?.remove();
      watchPosition.current = null;
    }
  };

  // location subscription
  useEffect(() => {
    (async () => {
      const { status } = await Location.requestForegroundPermissionsAsync();
      if (status !== "granted") {
        setManualGpsCoordinates(true);
        setPlaceholderText(Placeholders.GPS.NoPermissions);
        Logger.Warn(`Permission to access foreground location was ${status}.`);
        return;
      }

      try {
        watchPosition.current = await Location.watchPositionAsync(
          {
            accuracy: Location.Accuracy.High,
            timeInterval: 5000,
          },
          ({ coords }) =>
            updateGpsCoordinateString(
              `${clipCoordinate(coords.latitude)}, ${clipCoordinate(
                coords.longitude
              )}`
            )
        );
      } catch (error) {
        setManualGpsCoordinates(true);
        setPlaceholderText(Placeholders.GPS.NoLocation);
        Logger.Error(`Location.watchPositionAsync() failed: ${error}`);
      }
    })();

    return stopGps;
  }, []);

  useEffect(() => {
    gpsCoordinatesRef?.current?.focus();
  }, [manualGpsCoordinates]);

  const enableManualGpsCoordinates = () => {
    stopGps();
    updateGpsCoordinateString("");
    setManualGpsCoordinates(true);
    setPlaceholderText(Placeholders.GPS.EnterCoordinates);
  };

  const confirmManualGpsCoordinates = () => {
    if (showGpsWarning) {
      Alert.alert(
        "Enter GPS coordinates manually?",
        "This message can be disabled in Settings.",
        [
          {
            text: "Yes",
            onPress: () => enableManualGpsCoordinates(),
          },
          {
            text: "No",
            style: "cancel",
          },
        ]
      );
    } else {
      enableManualGpsCoordinates();
    }
  };

  const gpsFieldProps = {
    style: formInputStyles.optionInputText,
    editable: manualGpsCoordinates,
    maxLength: 30,
    placeholder: placeholderText,
    defaultValue: gpsCoordinateString,
    returnKeyType: "done" as ReturnKeyTypeOptions,
  };

  return (
    <>
      <Text style={formInputStyles.optionStaticText}>
        GPS Coordinates (latitude, longitude)
      </Text>
      {!manualGpsCoordinates && (
        /* user confirms they want to enter GPS coordinates manually */
        <Pressable
          onPress={() => confirmManualGpsCoordinates()}
          testID={TestIds.GPS.gpsManualPressableTestId}
        >
          <View pointerEvents="none">
            <TextInput {...gpsFieldProps} />
          </View>
        </Pressable>
      )}
      {manualGpsCoordinates && (
        <TextInput
          {...gpsFieldProps}
          testID={TestIds.GPS.gpsManualInputTestId}
          onChangeText={(value) => updateGpsCoordinateString(value)}
          onSubmitEditing={() => onSubmitEditing()}
          ref={gpsCoordinatesRef}
        />
      )}
    </>
  );
}

GpsCoordinatesInput.propTypes = {
  setGpsCoordinates: PropTypes.func.isRequired,
  onSubmitEditing: PropTypes.func.isRequired,
};

export default GpsCoordinatesInput;
