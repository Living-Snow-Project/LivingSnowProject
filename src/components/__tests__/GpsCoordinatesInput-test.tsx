import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import * as Location from "expo-location";
import GpsCoordinatesInput from "../forms/GpsCoordinatesInput";
import Placeholders from "../../constants/Strings";

const requestForegroundPermissionsAsyncSuccessMock = () =>
  jest
    .spyOn(Location, "requestForegroundPermissionsAsync")
    .mockImplementation(() =>
      Promise.resolve({
        status: "granted",
      } as Location.LocationPermissionResponse)
    );

const watchPositionAsyncSuccessMock = () =>
  jest
    .spyOn(Location, "watchPositionAsync")
    .mockImplementation((options, callback) => {
      const location = {
        coords: {
          latitude: 123.456,
          longitude: -98.765,
        },
      };

      callback(location as Location.LocationObject);

      const sub: Location.LocationSubscription = {
        remove() {},
      };

      return Promise.resolve<Location.LocationSubscription>(sub);
    });

describe("GpsCoordinatesInput tests", () => {
  const onSubmitEditing = jest.fn();
  const setGpsCoordinates = jest.fn();

  const requestForegroundPermissionsAsyncFailureMock = () => {
    jest
      .spyOn(Location, "requestForegroundPermissionsAsync")
      .mockImplementation(() =>
        Promise.resolve({
          status: "denied",
        } as Location.LocationPermissionResponse)
      );
  };

  const renderGpsCoordinatesInput = () =>
    render(
      <GpsCoordinatesInput
        onSubmitEditing={onSubmitEditing}
        setGpsCoordinates={setGpsCoordinates}
      />
    );

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("renders", async () => {
    watchPositionAsyncSuccessMock();
    requestForegroundPermissionsAsyncSuccessMock();

    const { getByDisplayValue, toJSON } = renderGpsCoordinatesInput();

    await waitFor(() => getByDisplayValue("123.456000, -98.765000"));
    expect(toJSON()).toMatchSnapshot();
  });

  test("permission denied", async () => {
    requestForegroundPermissionsAsyncFailureMock();

    const { getByPlaceholderText } = renderGpsCoordinatesInput();

    await waitFor(() => getByPlaceholderText(Placeholders.GPS.NoPermissions));
  });

  test("failure to watchPositionAsync", async () => {
    jest
      .spyOn(Location, "watchPositionAsync")
      .mockImplementation(() => Promise.reject(Error("mock failure")));
    requestForegroundPermissionsAsyncSuccessMock();

    const { getByPlaceholderText } = renderGpsCoordinatesInput();

    await waitFor(() => getByPlaceholderText(Placeholders.GPS.NoLocation));
  });

  test("extra parenthesis and spaces are clipped", async () => {
    requestForegroundPermissionsAsyncFailureMock();

    const setCoordinates = jest.fn();
    const { getByPlaceholderText } = render(
      <GpsCoordinatesInput
        onSubmitEditing={onSubmitEditing}
        setGpsCoordinates={setCoordinates}
      />
    );

    await waitFor(() => getByPlaceholderText(Placeholders.GPS.NoPermissions));

    fireEvent.changeText(
      getByPlaceholderText(Placeholders.GPS.NoPermissions),
      "(56.789  , 454.76)   "
    );

    expect(setCoordinates).toBeCalledWith("56.789", "454.76");
  });

  test("done editing", async () => {
    requestForegroundPermissionsAsyncFailureMock();

    const submitEditing = jest.fn();
    const { getByPlaceholderText } = render(
      <GpsCoordinatesInput
        onSubmitEditing={submitEditing}
        setGpsCoordinates={setGpsCoordinates}
      />
    );

    await waitFor(() => getByPlaceholderText(Placeholders.GPS.NoPermissions));
    fireEvent(
      getByPlaceholderText(Placeholders.GPS.NoPermissions),
      "onSubmitEditing"
    );
    expect(submitEditing).toHaveBeenCalled();
  });
});
