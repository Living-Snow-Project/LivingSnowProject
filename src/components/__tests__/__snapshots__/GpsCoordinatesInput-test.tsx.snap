// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GpsCoordinatesInput tests renders 1`] = `
Array [
  <Text
    style={
      Object {
        "fontSize": 15,
        "marginTop": 3,
      }
    }
  >
    GPS Coordinates (latitude, longitude)
  </Text>,
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="gps-manual-pressable"
  >
    <View
      pointerEvents="none"
    >
      <TextInput
        allowFontScaling={true}
        defaultValue="123.456000, -98.765000"
        editable={false}
        maxLength={30}
        placeholder="Looking for GPS signal"
        rejectResponderTermination={true}
        returnKeyType="done"
        style={
          Object {
            "backgroundColor": "#efefef",
            "borderColor": "gray",
            "borderRadius": 4,
            "borderWidth": 1,
            "minHeight": "8%",
            "paddingHorizontal": 15,
            "paddingVertical": 15,
          }
        }
        underlineColorAndroid="transparent"
      />
    </View>
  </View>,
]
`;
